kuzzleio/kuzzle-admin-console
# Refactor de la muerte

## What does this PR do ?
Something radically out of scope but that fixes a behavior everybody was complaining about.

Makes the AC properly handle the "Connecting...", "Connected", "Temporarily offline" states. Now, whenever Kuzzle goes down, the AC just shows a toast instead of changing the current page.

To implement this, a whole lot of stuff has been refactored. A whole lot of junky code has been cleaned up and it feels damn good.

### How should this be manually tested?
* Launch a Kuzzle instance
* Connect the AC to it
* Stop the Kuzzle instance
* Verify that the AC displays a toast
* Relaunch the Kuzzle instance
* Verify that the toast is gone

:warning: All the tests of the following specs must be :heavy_check_mark: 
* 404 page
* Collection management
* Document list
* Document update/replace
* Environments
* Indexes
* Login
* Search

### Other changes and Boyscout
- [x] Improve the handling of JSON parsing errors on environment load: a toast should notify that the environments are malformed and dump the content of the LocalStorage item to console.
- [x] Handle the case where no current environment is set (currently defaults to first available one)
- [x] Properly handle delete environments, both when still env left and when no env left.
- [x] Nice and smooth environment import with proper error messages
- [x] Nice and smooth entironment switch
- [x] Improve the handling of an environment update: if the URL of the Kuzzle server has changed, the connection should be reset.
- [x] Proper first admin check
- [x] Connection spinner on first page load
- [x] Token expired check
